include/aws/common/task_scheduler.h

```diff
@@ -0,0 +1,101 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/common.h>
+#include <aws/common/priority_queue.h>
+#include <stdint.h>
+
+typedef void *aws_task_arg;
+
+/**
+ * Pointer to the scheduled function
+ */
+typedef void(*aws_task_fn)(aws_task_arg arg);
+
+/**
+ * Monotonic clock function
+ */
+typedef int(*aws_task_scheduler_clock)(uint64_t *);
+
+struct aws_task {
+    aws_task_fn fn;
+    aws_task_arg arg;
+};
+
+struct aws_task_scheduler {
+    struct aws_allocator *alloc;
+    struct aws_priority_queue queue;
+    uint64_t owning_thread;
+    aws_task_scheduler_clock clock;
+};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    /**
+     * Initializes a task scheduler instance with a clock function and config. A priority queue will be allocated 
+     * in dynamic mode with initial size of 7
+     * The recommended clock functions are in <aws/common/clock.h>
+     *
+     * Note: This needs to be called from the thread that will own the scheduler. Cross-thread access is not allowed 
+     * and will always result in an error code being returned.
+     */
+    AWS_COMMON_API int aws_task_scheduler_init_default(struct aws_task_scheduler *scheduler, 
+            aws_task_scheduler_clock clock, struct aws_allocator *alloc);
+
+    /**
+     * Initializes a task scheduler instance with a priority queue, clock function and config.
+     * The recommended clock functions are in <aws/common/clock.h>
+     *
+     * Note: This needs to be called from the thread that will own the scheduler. Cross-thread access is not allowed 
+     * and will always result in an error code being returned.
+     */
+    AWS_COMMON_API void aws_task_scheduler_init_with_queue(struct aws_task_scheduler *scheduler, 
+            struct aws_priority_queue *queue, aws_task_scheduler_clock clock);
+
+    /**
+     * Cleans up any memory allocated, and prepares the instance for reuse or deletion.
+     */
+    AWS_COMMON_API void aws_task_scheduler_clean_up(struct aws_task_scheduler *scheduler);
+
+    /**
+     * Gets the next task that is ready for execution. Tasks in the future will not be returned, 
+     * even if they are the highest priority. next_run_time, if it is not null, and there are any tasks scheduled
+     * will be set to the timestamp for the highest priority task. This is a useful hint for setting timeouts on
+     * event loops, or thread sleeps. If no tasks are scheduled, this value will be set to 0. task is copied
```

Comment:
How do we indicate when no task is ready for execution?

Reply:
So between answering this question and fixing the bug you mentioned below (incorrect timestamp) this API has become too complicated for my taste. There're four possible outcomes to the two output parameters. The edge case is when there's only one task left in the queue, what should the value of the "next_run_time" be. Ideally return a different error code. But the implementation becomes too "branchy".

It's better to split those two functions. aws_task_scheduler_next_task() and aws_task_scheduler_next_run_time().

Thoughts?